### PR TITLE
Further simplify the flow of pdf text output.

### PR DESCRIPTION
**Pull request by anntzer**
## PR Summary

... and, as a side benefit, correctly kern between 1-byte and 2-byte
chunks.

followup to #14872.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/backends/backend_pdf.py [hunk 1]
```diff
@@ -2115,116 +2104,77 @@ def draw_text(self, gc, x, y, s, prop, angle, ismath=False, mtext=None):
...
-            """
-            chunks = [(a, ''.join(b)) for a, b in chunks]
-
+        # There is no way to access multibyte characters of Type 3 fonts,
+        # as they cannot have a CIDMap.  Therefore, in this case we break
+        # the string into chunks, where each chunk contains either a string
+        # of consecutive 1-byte characters or a single multibyte character.
+        # Each chunk is emitted with a separate command: 1-byte characters use
+        # the regular text show command (Tj), whereas multibyte characters
+        # use the use XObject command (Do).  (If using Type 42 fonts, all of
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_pdf.py]:
```suggestion
        # use the XObject command (Do).  (If using Type 42 fonts, all of
```

File: lib/matplotlib/backends/backend_pdf.py [hunk 2]
```diff
@@ -2115,116 +2104,77 @@ def draw_text(self, gc, x, y, s, prop, angle, ismath=False, mtext=None):
...
+                    s,
+                    _text_layout.layout(s, font, kern_mode=KERNING_UNFITTED)):
+                if ord(char) <= 255:
+                    if prev_was_singlebyte:
+                        singlebyte_chunks[-1][1].append(char)
+                    else:
+                        singlebyte_chunks.append((char_x, [char]))
+                    prev_was_singlebyte = True
+                else:
+                    multibyte_chunks.append((char_x, [char]))
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_pdf.py]: Why is the list needed around the character?

File: lib/matplotlib/backends/backend_pdf.py [hunk 3]
```diff
@@ -2115,116 +2104,77 @@ def draw_text(self, gc, x, y, s, prop, angle, ismath=False, mtext=None):
...
-                        self.file.output(Name(name), Op.use_xobject)
-                        self.file.output(Op.grestore)
-                newx = char_x + (  # ... including the last character's advance
-                    font.load_glyph(char_idx, flags=LOAD_NO_HINTING)
-                    .linearHoriAdvance / 65536)
-
+            # Then emit all the multibyte characters, one at a time.
+            for start_x, (char,) in multibyte_chunks:
+                glyph_name = font.get_glyph_name(
+                    font.get_char_index(ord(char)))
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_pdf.py]: You can't just put `char_idx` into `multibyte_chunks` instead?

File: lib/matplotlib/backends/backend_pdf.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/backends/backend_pdf.py]: fixed, thanks

File: lib/matplotlib/backends/backend_pdf.py (hunk 2 above)
Comment (anntzer) [reply lib/matplotlib/backends/backend_pdf.py]: I wanted to keep the format of singlebyte_chunks and multibyte_chunks the same but there isn't reall any benefit for that so renamed the latter to multibyte_glyphs and changed it.

File: lib/matplotlib/backends/backend_pdf.py (hunk 3 above)
Comment (anntzer) [reply lib/matplotlib/backends/backend_pdf.py]: same as https://github.com/matplotlib/matplotlib/pull/14886/files#r307636753, now changed

**Comment by anntzer** [comment]
Added a second commit renaming "char_idx" to "glyph_idx" throughout, as this patch is a bit less confusing with that (avoids confusion between unicode codepoint and glyph indices); fwiw freetype's doc appears to use "glyph index"... except that the function to retrieve them is FT_Get_Char_Index, but heh.

**Pull request merged**